### PR TITLE
Add profile to support spark 2.4.0

### PR DESCRIPTION
**Pull request by birdstorm**
Support Spark 2.4.0 by profile.

### CODE REVIEW

**Comment by birdstorm** [comment]
> please run format

done.

File: core/pom.xml [hunk 1]
```diff
@@ -305,7 +309,8 @@
                 <version>0.10_1.5.1</version>
                 <configuration>
                     <sourceDirectories> <!-- (Optional) Paths to source-directories. Overrides ${project.build.sourceDirectory} -->
-                        <param>${project.basedir}/src/main/scala</param>
+                        <sourceDirectory>${project.basedir}/src/main/scala</sourceDirectory>
+                        <sourceDirectory>${project.basedir}/src/main/${profile.source.directory}</sourceDirectory>
```
Comment (marsishandsome) [inline core/pom.xml]: shall we add both folder `spark-2.3` and `spark-2.4`?

Comment (birdstorm) [reply core/pom.xml]: I'm not sure of that. Both implementation is reasonable.

**Comment by marsishandsome** [comment]
in order to keep compatibility, there are three ways:
1. use different source code in compiling stage (what birdstorm did)
2. compile both version's target classes, but not generate in classpath (still in jar), according to spark's version at runtime, dynamically load the correct classes
3. according to spark's version at runtime, dynamically generate classes and load classes

File: core/pom.xml (hunk 1 above)
Comment (marsishandsome) [reply core/pom.xml]: we should format all the source code

**Comment by birdstorm** [comment]
> * use different source code in compiling stage (what birdstorm did)
> * compile both version's target classes, but not generate in classpath (still in jar), according to spark's version at runtime, dynamically load the correct classes
> * according to spark's version at runtime, dynamically generate classes and load classes

After discussion with @zanmato1984 , I believe by using scheme with profile is a better way to deal with multi-version dependency. We can use reflection that generates classes at runtime for minor versions -- Spark 2.3.x -- and combine with using different source roots for major versions -- Spark 2.x -- so that alternately we will introduce two releases for TiSpark, i.e., tispark_spark2.3 and tispark_spark2.4.

File: core/pom.xml (hunk 1 above)
Comment (birdstorm) [reply core/pom.xml]: so that we shall hardcode all available profiles, and remember to change them when new profiles are provided/modified.

**Comment by marsishandsome** [comment]
> After discussion with @zanmato1984 , I believe using scheme with profile is a better way to deal with multi-version dependency. We can use reflection that generates classes at runtime for minor versions -- Spark 2.3.x -- and combine with using different source roots for major versions -- Spark 2.x -- so that alternately we will introduce two releases for TiSpark, i.e., tispark_spark2.3 and tispark_spark2.4.

sounds good

File: core/pom.xml (hunk 1 above)
Comment (marsishandsome) [reply core/pom.xml]: or ci `/build` command will trigger both `-Pspark2.3` and `-Pspark2.4` profile

Comment (birdstorm) [reply core/pom.xml]: It is!

**Pull request merged**